src/hotspot/share/runtime/threadSMR.hpp

```diff
@@ -203,6 +205,8 @@ class ThreadsList : public CHeapObj<mtThread> {
   int find_index_of_JavaThread(JavaThread* target);
   JavaThread* find_JavaThread_from_java_tid(jlong java_tid) const;
   bool includes(const JavaThread * const p) const;
+
+  static bool is_valid(ThreadsList* list) { return list->_magic == THREADS_LIST_MAGIC; }
```

Comment:
This seems like a necessary, but not sufficient, check for validity. Unless we zap memory when these things are freed I don't see how this ensures validity.

Reply:
You missed L652 in the `ThreadsList::~ThreadsList()` above:

 _magic = 0xDEADBEEF;

When the destructor runs, the magic value is stomped. I've captured
failures where the is_valid() check observes 0xDEADBEEF and also
failures where the ThreadsList has been recycled into something else
and the `_magic` value is some other value that doesn't match
`THREADS_LIST_MAGIC`.

Reply:
Sorry yes I did miss it - I misread the diff due to hidden lines.